core/src/main/java/io/questdb/cairo/PlainRecordChain.java

```diff
@@ -0,0 +1,122 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo;
+
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.cairo.vm.Vm;
+import io.questdb.cairo.vm.api.MemoryARW;
+import io.questdb.std.MemoryTag;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * PlainRecordChain is similar to RecordChain, except that it stores the record's startOffset in a separate memory
+ * location instead of at the record header.
+ * This enhances its random access capability, making it behave more like an array,
+ * while sacrificing the ability to access records like a linked list.
+ */
+public class PlainRecordChain extends RecordChain {
+
+    private final MemoryARW recordsStartOffset;
+    private long size = 0L;
+    private long nextRecordIndex = 0L;
+
+    public PlainRecordChain(@NotNull ColumnTypes columnTypes, @NotNull RecordSink recordSink, long pageSize, int maxPages) {
+        super(columnTypes, recordSink, pageSize, maxPages);
+        this.recordsStartOffset = Vm.getCARWInstance(pageSize >> 3, maxPages, MemoryTag.NATIVE_RECORD_CHAIN);
+    }
+
+    public long put(Record record) {
+        long offset = beginRecord();
+        recordSink.copy(record, this);
+        return offset;
+    }
+
+    public long beginRecord() {
+        recordOffset = varAppendOffset;
+        recordsStartOffset.putLong(recordOffset);
+        size++;
+        mem.jumpTo(recordOffset + varOffset);
+        varAppendOffset = recordOffset + varOffset + fixOffset;
+        return recordOffset;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (nextRecordIndex < size) {
+            final long offset = recordsStartOffset.getLong(nextRecordIndex * 8);
+            recordA.of(offset);
+            nextRecordIndex++;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void toTop() {
+        nextRecordIndex = 0;
+    }
+
+    public void toBottom() {
+        nextRecordIndex = size - 1;
+    }
+
+    public boolean hasPrev() {
+        if (nextRecordIndex >= 0) {
+            final long offset = recordsStartOffset.getLong(nextRecordIndex * 8);
+            recordA.of(offset);
+            nextRecordIndex--;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void calculateSize(SqlExecutionCircuitBreaker circuitBreaker, Counter counter) {
+        counter.add(size);
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        size = 0L;
+        recordsStartOffset.close();
+    }
+
+    @Override
+    protected long rowToDataOffset(long row) {
+        return row;
+    }
+
+    @Override
+    protected RecordChainRecord newChainRecord() {
+        return new RecordChainRecord(columnCount) {
```

Comment:
it is better not to have this as anonymous class. In Java, new class will be created on each invocation, which is sub-optimal. That is in addition to object instance.